### PR TITLE
Make float::from_bits transmute

### PR DESCRIPTION
**Pull request by Gankra**
See commit message for details.

See also this discussion here: https://github.com/rust-lang/rust/issues/40470#issuecomment-343803381

(may require libs team discussion before merging)

### CODE REVIEW

**Comment by Gankro** [comment]
It's possible this is also valid on powerpc/sparc64/asmjs but I'm keeping this down to the platforms where I know this works and I actually care about. :)

**Comment by sfackler** [comment]
I'm definitely on board with this, but I think we should get everyone's eyes on it.

@rfcbot fcp merge

File: src/libstd/f32.rs [hunk 1]
```diff
@@ -1044,17 +1044,24 @@ impl f32 {
...
-        if v & EXP_MASK == EXP_MASK && v & FRACT_MASK != 0 {
-            // While IEEE 754-2008 specifies encodings for quiet NaNs
-            // and signaling ones, certain MIPS and PA-RISC
-            // CPUs treat signaling NaNs differently.
-            // Therefore to be safe, we pass a known quiet NaN
-            // if v is any kind of NaN.
-            // The check above only assumes IEEE 754-1985 to be
-            // valid.
-            v = unsafe { ::mem::transmute(NAN) };
+        if cfg!(not(any(
```
Comment (sfackler) [inline src/libstd/f32.rs]: Can you add comments here and below noting why we only do this on some platforms?

Comment (Gankro) [reply src/libstd/f32.rs]: The existing comment seems to address it, no?

Comment (sfackler) [reply src/libstd/f32.rs]: The existing comment just says why we reset to a "canonical" NaN instead of just masking off the signal bit.

Comment (Gankro) [reply src/libstd/f32.rs]: "While IEEE 754-2008 specifies encodings for quiet NaNs and signaling ones, certain MIPS and PA-RISC CPUs treat signaling NaNs differently."

Comment (Gankro) [reply src/libstd/f32.rs]: Err wait, I think I don't follow what you're asking/saying?

Comment (sfackler) [reply src/libstd/f32.rs]: The documentation to this function states that "due to the implications of Rust's safety promises being uncertain, if the representation of a signaling NaN "sNaN" float is passed to the function, the implementation is allowed to return a quiet NaN instead".

Someone looking at the implementation may wonder why those questions about safety promises are not a problem on these four architectures, but are on others. There should be a comment saying why this adjustment isn't needed.

Comment (Gankro) [reply src/libstd/f32.rs]: I would instead suggest changing the documentation to be "due to some platforms not conforming to IEEE 758-2008..."

Comment (sfackler) [reply src/libstd/f32.rs]: I don't understand what IEEE 758-2008 compliance has to do with LLVM treating sNaN as UB or not.

**Comment by Gankro** [comment]
I have changed the PR to transmute on all platforms. The platform-specific cfg was based on me getting confused as a result of the implementation, documentation, and example all being inconsistent with eachother. See commit message for details.

**Comment by rkruppe** [comment]
Should the weasel words around sNaN be removed, i.e., should we commit to not masking sNaN?

And if so, should the the docs also include the note about non-IEEE platforms that's in the commit, to warn people against using it for platform-independent de-/serialization?

**Comment by Gankro** [comment]
Note that the platforms in question are IEEE-1985 compliant. It's just how to interpret the signaling bit wasn't spec'd until the 2008 version. Upon reflection, this implementation is maximally good at network interop if you're doing NaN-boxing that just slurps up the mantissa bits. It's not as good if you're specifically just caring about whether the NaN signals or not.

The alternative would be to preserve signaling-ness, at the cost of clobbering the signaling bit across platforms.

On balance my submitted implementation seems the most useful, and if the alternative behaviour is desired, then it should be a seperate function (since this behaviour is still desirable).

In practice, I expect no one will notice or care.

**Comment by sfackler** [comment]
I think the guarantees we want to document are that:

* Non-NaN values will always round trip exactly.
* NaN values will remain NaN after a round trip between any set of architectures.

Does that seem right?

**Comment by Gankro** [comment]
That is a reasonably minimal specification that I would be happy with, yes. (pushed test changes)

**Comment by sfackler** [comment]
> All values will always round trip exactly between the same architecture.

I think we also want to guarantee this, right?

**Comment by BurntSushi** [comment]
@Gankro Thanks for all the leg work on this. :-) I admit, I still feel a little squishy over this, but folks with more experience than me on these matters seem quite confident, so I'm content to move forward.

**Comment by alexcrichton** [comment]
I think the docs may want to be updated here as well, but could we perhaps word them in such a way that implies that despite the current behavior we may change things in the future if necessary? (in that do we really trust LLVM to never change in this regard?)

**Comment by Gankro** [comment]
Willing to change docs if desired (though they are still *technically* correct), but I'd like to get sign off from the rest of the libs team before digging into that. (the impl should be ~independent)

**Comment by sfackler** [comment]
We discussed this at the last libs meeting and decided that the guarantees in https://github.com/rust-lang/rust/pull/46012#issuecomment-344770349 seem reasonable.

**Comment by Gankro** [comment]
After some discussion on IRC, it seems like it would be better to simply state "it's just transmute", and explain the implications of this. This ends up providing basically the same guarantees as the libs team agreed to (except guaranteeing that cross-platform transfer is in fact bit-exact).

Pushed updated docs.

File: src/libstd/f32.rs [hunk 2]
```diff
@@ -1018,17 +1021,33 @@ impl f32 {
...
+    /// It turns out this is incredibly portable, for two reasons:
+    ///
+    /// * Floats and Ints have the same endianess on all supported platforms.
+    /// * IEEE-754 very precisely specifies the bit layout of floats.
+    ///
+    /// However there is one caveat: prior to the 2008 version of IEEE-754, how
+    /// to interpret the NaN signaling bit wasn't actually specified. Most platforms
+    /// (notably x86 and ARM) picked the interpretation that was ultimately 
+    /// standardized in 2008, but some didn't (notably MIPS). As a result, all
+    /// signaling NaNs on MIPS are quiet NaNs on x86, and vice-versa.
```
Comment (sfackler) [inline src/libstd/f32.rs]: Is this really the case, or is it that the signalingness is selected by some payload bit in the other encoding?

Comment (Gankro) [reply src/libstd/f32.rs]: They literally agree on the bit, but one says 1=signaling, and the other says 0=signaling.

Comment (sfackler) [reply src/libstd/f32.rs]: Oh, lol

**Comment by sfackler** [comment]
Build's red, but r=me otherwise.

**Comment by Gankro** [comment]
Fixed? (tidy whitespace)

**Comment by kennytm** [comment]
@bors retry 

The AppVeyor `dist i686-pc-windows-msvc` job was "cancelled by user" with [this message](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5476/job/p8tp4w7bklsv7gpl/messages) but the failure is not notified to GitHub:

```
Timeout performing HVALS account_clouds_0, inst: 1, mgr: PrepareActiveSockets, err: never, queue: 4, qu: 0, qs: 2, qc: 2, wr: 0, wq: 0, in: 0, ar: 0, clientName: AppVeyor-Appveyor.Worker_IN_1, IOCP: (Busy=1,Free=999,Min=200,Max=1000), WORKER: (Busy=24,Free=32743,Min=200,Max=32767), Local-CPU: 4.09% (Please take a look at this article for some common client-side issues that can cause timeouts: https://github.com/StackExchange/StackExchange.Redis/tree/master/Docs/Timeouts.md)
```

**Comment by Gankro** [comment]
The Curse Is Lifted

**Pull request merged**